redbeat/schedulers.py

```diff
@@ -320,18 +325,22 @@ def schedule(self):
         client = redis(self.app)
 
         with client.pipeline() as pipe:
-            pipe.zrangebyscore(self.app.redbeat_conf.schedule_key, 0, max_due_at)
+            pipe.zrangebyscore(self.app.redbeat_conf.schedule_key, -1, max_due_at, withscores=True)
 
             # peek into the next tick to accuratly calculate sleep between ticks
             pipe.zrangebyscore(self.app.redbeat_conf.schedule_key,
                                '({}'.format(max_due_at),
                                max_due_at + self.max_interval,
-                               start=0, num=1)
+                               start=0, num=1, withscores=True)
             due_tasks, maybe_due = pipe.execute()
 
         logger.info('Loading %d tasks', len(due_tasks) + len(maybe_due))
         d = {}
-        for key in due_tasks + maybe_due:
+        for key, score in due_tasks + maybe_due:
+            if score < 0:
+                logger.info('Scheduler: Removing ended schedule %s', key)
+                client.zrem(self.app.redbeat_conf.schedule_key, key)
```

Comment:
@sibson Thinking about this more, it seems a bit strange to make `from_key` raise an error - the schedule having no more occurrences doesn't mean it's an invalid ScheduleEntry, just that the scheduler doesn't have to worry about checking it anymore. I can leave the main loop untouched, and setting the score of ended schedules to `-1` means that they will be ignored on each tick since your `zrangebyscore` starts at 0. Does that work?